File: Two-Pointer/findClosestElements.py
from typing import List
import sys

class Solution:
    def findClosestElements(self, arr: List[int], k: int, x: int) -> List[int]:
        n = len(arr)
        minDiff = sys.maxsize
        minIndex = -1
        left = []
        right = []

        for i,num in enumerate(arr):
            if abs(num - x) < minDiff:
                minIndex = i
                minDiff = abs(num - x)

        i = minIndex - 1
        j = minIndex + 1
        k = k - 1

        while k > 0:
            if i >= 0 and (j >= n or abs(x - arr[i]) <= abs(x - arr[j])):
                left.append(arr[i])
                i -= 1
            elif j < n and (i < 0 or abs(x - arr[j]) < abs(x - arr[i])):
                right.append(arr[j])
                j += 1
            k -= 1


        result = []
        left.reverse()

        for val in left:
            result.append(val)
        result.append(arr[minIndex])
        for val in right:
            result.append(val)

        return result




        

File: Sliding-Window/findClosestElements.js
function findClosestElements(arr, k, x) {
  let index = 0;
  let minDiff = Number.MAX_SAFE_INTEGER;

  for (let i = 0; i < arr.length; i++) {
    if (Math.abs(x - arr[i]) < minDiff) {
      minDiff = Math.abs(x - arr[i]);
      index = i;
    }
  }

  const left = [];
  const right = [];
  let i = index - 1;
  let j = index + 1;

  right.push(arr[index]);
  k--;

  while (k > 0 && i >= 0 && j < arr.length) {
    if (Math.abs(arr[i] - x) < Math.abs(arr[j] - x)) {
      left.push(arr[i]);
      i--;
      k--;
    } else {
      right.push(arr[j]);
      j++;
      k--;
    }
  }

  while (k > 0 && i >= 0) {
    left.push(arr[i]);
    i--;
    k--;
  }

  while (k > 0 && j < arr.length) {
    right.push(arr[j]);
    j++;
    k--;
  }

  return [...left.reverse(), ...right];
}
